exonum/src/helpers/fabric/details.rs

```diff
@@ -392,6 +406,47 @@ impl GenerateNodeConfig {
 
         (external_address, listen_address)
     }
+
+    /// * `key` - expects EXONUM_SERVICE_PASS or EXONUM_CONSENSUS_PASS.
+    fn get_passphrase(context: &Context, key: &str) -> Vec<u8> {
+        context
+            .get_flag_occurrences(NO_PASSWORD)
+            .map(|_| "".to_string())
+            .or_else(|| env::var(key).ok())
+            .or_else(|| {
+                let stdin = io::stdin();
+                let stdin_locked = stdin.lock();
+                Self::prompt_passphrase(stdin_locked, io::stdout()).ok()
+            }).unwrap()
+            .into_bytes()
+    }
+
+    fn prompt_passphrase<R: BufRead, W: Write>(mut reader: R, mut writer: W) -> io::Result<String> {
+        loop {
+            write!(&mut writer, "Enter passphrase (empty for no passphrase): ").unwrap();
+            let password = rpassword::read_password_with_reader(Some(&mut reader))?;
```

Comment:
I'd suggest using `rpassword::read_password_from_tty` instead of explicitly manipulating stdin and stdout.
It is more correct to read passwords directly from terminal in UNIX world.

I guess you have used `BufRead` exactly to be able to write unit-tests for this function. However, I'm not sure that we need to test this behavior, retrying not that complex and we can trust `rpassword` to do its thing correctly. This test seems to provide little value. It would be more useful to have an integration test which verifies environment variable handling together with that `NO_PASSWORD` thing.

Also, I'm curious: how does this behave if there is no interactive terminal? E.g., when Exonum node is launched as a service with stdin and stdout bound to `/dev/null`?

Reply:
> This test seems to provide little value.

I think it's important to test our prompt behavior: what exactly is displayed to used and under what conditions. I can't see a way to do it in our integration tests.

> Also, I'm curious: how does this behave if there is no interactive terminal? E.g., when Exonum node is launched as a service with stdin and stdout bound to `/dev/null`?

You can always pass the password with environment variables, so you won't be prompted to enter the password manually.

Reply:
> You can always pass the password with environment variables, so you won't be prompted to enter the password manually.

But what if I forget to set the environment variable? Will Exonum node hang indefinitely, assume an empty password, or prompt the user on some random terminal for the password?